### ISSUE TITLE
how to do integration test combine the oidc of IdentityServer , MVC client and WebAPI

### ISSUE THREAD

**Issue by jingliancui** [dotnet-architecture/eShopOnContainers#898]
Now I have a project in my job to develop
So I simulate eshopOnContainer's structure for my project(3 project now):
Asp.net core webapi project for the resource of the job's project
Asp.net mvc project for client to serve
Identity server 4 project use asp.net identity with oidc feature to protect the webapi project.

All config is done and run fine in the debug and release mode.
Base on it, I start the integration test projects.
Issue show In the oidc scenen:
1. MVC client have a action for example map http://localhost:10001/home/Test, this action with [authorize] attribute.When unlogin user to this link , it will redirect user to the idserver login page to login.
2. After IdServer finished logined operation.It redirect user to the MVC client's http://localhost:10001/home/Test ,and can be retrieve these resource fine.

Above  scene I try to do it in the integration test environment , but It's difficult for use the `WebApplicationFactory.CreateClient()`'s `httpclient` to retrieve 2 micro service (one is mvcclient and one is idserver).Because test environment just run in the in-memory , and I have no idea config the httpclient have ability to access 2 micro service (1 client access 2 testserver).

I want know if my direct is wrong or not.
Or can you give me some suggestion for me to do the integration test simulated the user's operation round tirp ?
At last , Why I want this because I have do the authorization in the oidc with asp.net identity (role base about),so if I use clientid and clientsecure to get the accesstoken,it's not have the role base about permission to access the protected api which protected by the role.
Result is this a challenge for these issue .

**Comment by mvelosop** [comment]
Hi @jingliancui, perhaps you should redirect your question to StackOverflow as this repo is focused on eShopOnContainer issues only.

I will leave the issue open for a while, just in case some one from the community can help you.

In the mean time, I suggest you to read the tests readme as you might find some direction there: https://github.com/dotnet-architecture/eShopOnContainers/tree/dev/test.

Hope this helps.

**Comment by jingliancui** [comment]
@mvelosop Hi ,Thanks for open this issue for wait people to discuss this architecture plomlem.
And i had post the same issue to Stack Overflow [here](https://stackoverflow.com/questions/54102820/how-to-do-integration-test-combine-the-oidc-of-identityserver-mvc-client-and-w)

Then I go to read the code you provide , and i found that you use the `AutoAuthorizeMiddleware` for test , I wank know if it can solve my issue scene. Or can you explain the middleware ?

**Comment by mvelosop** [comment]
@jingliancui, without getting too deep in the details, I think you might find this some guidance here: https://github.com/dotnet-architecture/eShopOnContainers/blob/dev/test/ServicesTests/Application.FunctionalTests/Services/IntegrationEventsScenarios.cs.

Hope this helps.

**Comment by jingliancui** [comment]
@mvelosop Hi,Yep,but I check the code you provide in IntegrationEventsScenarios.cs , `var basketServer = new BasketScenariosBase().CreateServer()` invoke the `UseStartup<BasketTestsStartup>` , then BasketTestsStartup have these:

```
        protected override void ConfigureAuth(IApplicationBuilder app)
        {
            if (Configuration["isTest"] == bool.TrueString.ToLowerInvariant())
            {
                app.UseMiddleware<AutoAuthorizeMiddleware>();
            }
            else
            {
                base.ConfigureAuth(app);
            }
        }
```
Then, AutoAuthorizeMiddleware still here........

**Comment by jingliancui** [comment]
You are right 😄 I think i shoud explore this project for me. Thanks so much
